### PR TITLE
BatchedMesh: add support for frustum culling per batched geometry

### PR DESCRIPTION
**Pull request by gkjohnson**
Related issue: #22376, #27111 

**Description**

Add support for frustum culling to BatchedMesh by storing the box and sphere bounds per geometry.

**Upcoming PRs**
- Raycasting, sorting support
- Object-level bounds
- Add support to ObjectLoader
- Move to core
- Documentation

### CODE REVIEW

**Comment by Mugen87** [comment]
So the PR adds a frustum culling check for each geometry of the batched mesh in `onBeforeRender()`.

This is a different strategy to what is implemented in other classes. Normally, I would have expected bounding volumes on batched mesh level (similar to `InstancedMesh` or `SkinedMesh`). The computation of these bounding volumes happens automatically by the renderer (or or other components like in context of raycasting) when the access happens for the first time (meaning when their values are still `null`). Whenever the bounds require an updated, `computeBoundingSphere()` or `computeBoundingBox()` needs to be called on _app level_ since the app updates the geometry data.

I think it would be more consistent to implement the bounding volume computation according to this policy for batched mesh as well.

**Comment by gkjohnson** [comment]
To be clear you're suggesting that we only generate bounding volumes on an as-needed basis for BatchedMesh, correct? I can adjust that if that's correct.

> Whenever the bounds require an updated, computeBoundingSphere() or computeBoundingBox() needs to be called on app level since the app updates the geometry data.

This case is a bit different, though, since we're talking about a geometry that could have hundreds or thousands of bounding boxes - only one of them could be invalidated. I'm thinking adding an optional argument to "setGeometryAt" that indicates that the bounds should be invalidated is a good option.

**Comment by Mugen87** [comment]
>  on an as-needed basis for BatchedMesh, correct?

Yes. 

> This case is a bit different, though, since we're talking about a geometry that could have hundreds or thousands of bounding boxes - only one of them could be invalidated. I'm thinking adding an optional argument to "setGeometryAt" that indicates that the bounds should be invalidated is a good option.

`BatchedMesh` is indeed special compared to other 3D objects but I think it would still be good to have bounds on object level so `projectObject()` in WebGL renderer can sort out the entire batched mesh if possible.

If you think it's important to have bounding volumes for each geometry of the batched mesh then I'm fine with that. But it would be good to add `BachtedMesh.boundingSphere` and `BachtedMesh.boundingBox` which enclose the entire batched mesh. Both volumes should be computed with `BachtedMesh.computeBoundingSphere()` and BachtedMesh.computeBoundingBox()`.

Ideally, frustum culling and raycasting test with the object level bounding volumes first, and then with the ones on geometry level.

**Comment by gkjohnson** [comment]
> BatchedMesh is indeed special compared to other 3D objects but I think it would be good to have bounds on object level so projectObject() in WebGL renderer can sort out the entire batched mesh if possible.

I don't know if I agree that this is so useful - but if this is something we want then I'd want the ability to disable that feature since keeping the full bounding box up to date with many dynamic objects spread throughout the scene is not super useful. In this case how do you feel about a different field like "perObjectFrustumCulled" or "batchFrustumCulling" that can be set to toggle the the "onBeforeRender" frustum culling feature?

Also I just updated the PR to generate the boxes and spheres as-needed. When a geometry is replaced the existing bounding box is used or it's marked as needing to be regenerated.

**Comment by Mugen87** [comment]
> In this case how do you feel about a different field like "perObjectFrustumCulled" or "batchFrustumCulling" that can be set to toggle the the "onBeforeRender" frustum culling feature?

That sounds good to me! However, in order to keep the class consistent I would be glad to see the object level features mentioned in https://github.com/mrdoob/three.js/pull/27120#issuecomment-1794451420. 

It's right that bounding volumes which enclose the entire batched mesh might not be useful for stuff like tiles rendering. But other scenarios with a smaller scale might benefit from sorting out the entire batched mesh when rendering. By implementing this approach, we keep the class consistent with `InstancedMesh` but can also provide more fine granular control over frustum culling with an additional property. `perObjectFrustumCulled` sounds good to me!

**Comment by gkjohnson** [comment]
Okay! I've just added "perObjectFrustumCulled" flag.

> That sounds good to me! However, in order to keep the class consistent I would be glad to see the object level features mentioned in https://github.com/mrdoob/three.js/pull/27120#issuecomment-1794451420.

I'll add this in another PR, then.

I guess one concern I have for things like raycasting or frustum culling when elements are dynamic (meaning the bounding volumes have to be regenerated every frame) is that the cost of regenerating the bounding box and / or sphere is more expensive than just checking the ray / frustum intersection with the child bounding boxes. Instancing would have the same issue, though I guess this would have to be tested.

**Comment by gkjohnson** [comment]
@Mugen87 I just updated the copy and toJSON functions to add support for bounds and perObjectFrustumCulled flag. This should be ready!

**Review by Mugen87 — APPROVED** [review]

File: examples/jsm/objects/BatchedMesh.js [hunk 1]
```diff
@@ -611,25 +756,55 @@ class BatchedMesh extends Mesh {
...
+
+		// prepare the frustum
+		if ( perObjectFrustumCulled ) {
+
+			_projScreenMatrix
+				.multiplyMatrices( camera.projectionMatrix, camera.matrixWorldInverse )
+				.multiply( this.matrixWorld );
+			_frustum.setFromProjectionMatrix(
+				_projScreenMatrix,
+				_renderer.isWebGPURenderer ? WebGPUCoordinateSystem : WebGLCoordinateSystem
```
Comment (LeviPesin) [inline examples/jsm/objects/BatchedMesh.js:769]: I think something like `_renderer.backend !== undefined ? _renderer.backend.coordinateSystem : WebGLCoordinateSystem` would be more correct. /ping @gkjohnson

Comment (gkjohnson) [reply examples/jsm/objects/BatchedMesh.js:769]: I'm not familiar enough with WebGPURenderer to know

Comment (LeviPesin) [reply examples/jsm/objects/BatchedMesh.js:769]: @sunag What do you think?

**Comment by sunag** [comment]
I think we can use `_renderer.coordinateSystem`, it is compatible with both renderers.

File: examples/jsm/objects/BatchedMesh.js (hunk 1 above)
Comment (sunag) [reply examples/jsm/objects/BatchedMesh.js:769]: I think we can use _renderer.coordinateSystem, it is compatible with both renderers.

**Pull request merged**